homeassistant/components/volumio/config_flow.py

```diff
@@ -0,0 +1,122 @@
+"""Config flow for Volumio integration."""
+import logging
+from typing import Optional
+
+from pyvolumio import CannotConnectError, Volumio
+import voluptuous as vol
+
+from homeassistant import config_entries, exceptions
+from homeassistant.const import CONF_HOST, CONF_ID, CONF_NAME, CONF_PORT
+from homeassistant.core import callback
+from homeassistant.helpers.aiohttp_client import async_get_clientsession
+from homeassistant.helpers.typing import DiscoveryInfoType
+
+from .const import DOMAIN  # pylint:disable=unused-import
+
+_LOGGER = logging.getLogger(__name__)
+
+
+DATA_SCHEMA = vol.Schema(
+    {vol.Required(CONF_HOST): str, vol.Required(CONF_PORT, default=3000): int}
+)
+
+
+async def validate_input(hass, host, port):
+    """Validate the user input allows us to connect."""
+    volumio = Volumio(host, port, async_get_clientsession(hass))
+
+    try:
+        return await volumio.get_system_info()
+    except CannotConnectError as error:
+        raise CannotConnect from error
+
+
+class ConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
+    """Handle a config flow for Volumio."""
+
+    VERSION = 1
+    CONNECTION_CLASS = config_entries.CONN_CLASS_LOCAL_POLL
+
+    def __init__(self):
+        """Initialize flow."""
+        self._host: Optional[str] = None
+        self._port: Optional[int] = None
+        self._name: Optional[str] = None
+        self._uuid: Optional[str] = None
+
+    @callback
+    def _async_get_entry(self):
+        return self.async_create_entry(
+            title=self._name,
+            data={
+                CONF_NAME: self._name,
+                CONF_HOST: self._host,
+                CONF_PORT: self._port,
+                CONF_ID: self._uuid,
+            },
+        )
+
+    async def _set_uid_and_abort(self):
+        await self.async_set_unique_id(self._uuid)
+        self._abort_if_unique_id_configured(
+            updates={
+                CONF_HOST: self._host,
+                CONF_PORT: self._port,
+                CONF_NAME: self._name,
+            }
+        )
+
+    async def async_step_user(self, user_input=None):
+        """Handle the initial step."""
+        errors = {}
+        if user_input is not None:
+            info = None
+            try:
+                self._host = user_input[CONF_HOST]
+                self._port = user_input[CONF_PORT]
+                info = await validate_input(self.hass, self._host, self._port)
+            except CannotConnect:
+                errors["base"] = "cannot_connect"
+            except Exception:  # pylint: disable=broad-except
+                _LOGGER.exception("Unexpected exception")
+                errors["base"] = "unknown"
+
+            if info is not None:
+                self._name = info.get("name", self._host)
+                self._uuid = info.get("id", None)
+                if self._uuid is not None:
```

Comment:
So UUID can be missing?
If so, is it guaranteed to be available in the zeroconf discovery? As the zerconf step assumes it is always there.

Reply:
UUID is always there in discovery. Without discovery you have to query Volumio for the UUID, but older versions (i.e. everything but the upcoming version :) ) don't implement that endpoint.